### ISSUE TITLE
Cannot build with current OCAML

### ISSUE THREAD

**Issue by alerque** [patoline/patoline#30]
I have been trying to spin up some packaging of Patoline via Arch Linux and/or Nix (with the ultimate goal of playing around with it in comparison with other typesetters *vis-à-vis* https://polytype.dev) but have been stymied so far. The current road block is that both of these distros use new OCAML tooling (ocaml 5.1.0 and dune 3.11.1), and it seems Patoline will only work with older OCAML versions.

At least one reason for this is that it depends on `early.ocaml` which was removed (or rather made private) in the Earley v3 release. The v2 release had it, but that version will not build with OCAML 5.

Is it possible to update the use of Earley and make any other adjustments needed so that Patoline can be built with OCAML 5, or is this a dead end?

**Comment by alerque** [comment]
[This comment over on Earley](https://github.com/rlepigre/ocaml-earley/issues/7#issuecomment-1871045844) suggests (at least to my reading) that the current state of Patoline is known to be broken but fixing it isn't trivial and requires reworking the parser to something that builds in current OCAML.

**Comment by rlepigre** [comment]
Yeah, that is mostly right. I think we could fix it, but the parsing situation is definitely suboptimal and hard to maintain (mainly due to bad support on the OCaml side, and us having to ship an OCaml parser).

If you plan to try Patoline, the easiest way is to rely on a specific opam switch. For instance, I would do something like the following after having installed opam via your distribution or following (these instructions)[https://opam.ocaml.org/doc/Install.html]:
```
mkdir patoline_experiments; cd patoline_experiments
opam switch create --empty .
eval $(opam env)
opam install patoline # had to retry this once for some reason
```
After doing this, you should be able to use patoline within the `patoline_experiments` folder (provided you run `eval $(opam env)` in any new shell you open).

For example, in the same shell, you can then do:
```
mkdir test1; cd test1
echo "Hello, world\!" > test.txp
patoline test.txp
```

**Comment by alerque** [comment]
Hmm, that actually might be workable to spin up to enable Patoline samples on Polytype. It takes a while to build and lands almost exactly 1Gig in the local env, but it does work on a modern system and it looks like we can probably cache the build part relatively easily. It also seems fairly deterministic/reproducible.
